### ISSUE TITLE
Leviathan: ruby proc during lightning phase triggers unintended failure

### ISSUE THREAD

**Issue by IdNkI** [DominickCobb-rs/DT2-Boss-Failure#1]
ruby proc during lightning phase triggers unintended failure
not extensively tested

**Comment by DominickCobb-rs** [comment]
Thank you for your commentary about it not being extensively tested, and for bringing this issue to my attention. You're welcome to edit code and submit pull requests to fix bugs and issues if you feel you're capable of writing and testing!

**Comment by DominickCobb-rs** [comment]
And to add to this: The ruby doesn't fail anywhere outside of the special attacks, as was intended and considered during development. Leviathan's special attacks are difficult to track and as a result the rubies were ignored for that. I'm sure I'll get around to it at some point, but otherwise you're welcome to peruse the [Leviathan](https://github.com/DominickCobb-rs/DT2-Boss-Failure/blob/master/src/main/java/com/dt2PerfectBossFailure/bosses/Leviathan.java) file and see how it works and what could be done differently.

**Comment by DominickCobb-rs** [comment]
If you want to help without coding, can you confirm whether the particle effect happens on Leviathan like the gif on the [wiki](https://oldschool.runescape.wiki/w/Ruby_bolts_(e))?

**Comment by IdNkI** [comment]
There was no particle effect of the ruby proc, I assume this is because the lightning overrides the particle effect?
Maybe a clip of the bug occurring might help you more so I attached a mp4.
The failure alert was: "You were hit by lightning."
Similar unintended failure happens when drinking divine potions during specials.

https://github.com/DominickCobb-rs/DT2-Boss-Failure/assets/154762792/5ae0575c-cf09-45d0-8681-5995f5f89a5e

I don't see an elegant way of fixing this issue but it might be possible to use the sound effect which is also used by ruby-counter. beware this won't work for players that have sounds completely off. (https://github.com/hyppis1/BoltProcCounter/blob/master/src/main/java/com/BoltProcCounter/BoltProcCounterPlugin.java)

**Comment by DominickCobb-rs** [comment]
Yes that would work, but creates an event bus race condition where the we have to determine if the sound effect of the ruby, which isn't always active, is played before or after the hitsplat is applied. I'd rather not implement something that works for 50% of players too. I need to figure out a way to handle the smoke blast and lightning in such a way that the ruby hitsplat is unaffected.

**Comment by anguspaterson-code** [comment]
Having a look over this, you could potentially use the following (I'm not a confident person) ;)

> `// Constants
    private static final String LEVIATHAN = "The Leviathan";
    public static final int[] LEVIATHAN_IDS = {NpcID.THE_LEVIATHAN, NpcID.THE_LEVIATHAN_12215, NpcID.THE_LEVIATHAN_12219, NpcID.THE_LEVIATHAN_12221};
    private static final int LEVIATHAN_RANGED_SPOT_ANIM = 2490;
    private static final int LEVIATHAN_MAGE_SPOT_ANIM = 2492;
    private static final int LEVIATHAN_MELEE_SPOT_ANIM = 2491;
    private static final int[] LEVIATHAN_SPECIAL_ATTACK_ANIMATIONS = {10285, 10286, 10287, 10289, 10288, 10290};
    private static final int[] LEVIATHAN_SMOKE_BLAST_BOULDERS_ANIM = {10289, 10288, 10287};
    private static final int LEVIATHAN_SMOKE_BLAST_ANIM = 10290;
    private static final int[] LEVIATHAN_LIGHTNING_ANIM = {10285, 10286};
    private static final int BOULDER_MOVE_ANIM = 1114;
    private static final int[] LEVIATHAN_BOULDER_SHADOWS = {2475, 2476, 2477, 2478, 2479, 2480};
    private int leviathanSpecial = -1;
    private static final int LEVIATHAN_REGION_ID = 8291;

    // Placeholders for IDs (replace with actual IDs using DevTools)
    private static final int DIVINE_POTION_VARBIT_ID = 1234;
    private static final int BLOOD_FORFEIT_SOUND_ID = 5678;
    private static final int BLOOD_FORFEIT_GRAPHIC_ID = 91011;

    private boolean inLeviathanRegion()
    {
        if (!client.getGameState().equals(GameState.LOGGED_IN) || client.getLocalPlayer() == null)
        {
            return false;
        }
        return plugin.getCurrentRegion() == LEVIATHAN_REGION_ID;
    }

    @Subscribe
    public void onHitsplatApplied(HitsplatApplied hitsplatApplied)
    {
        if (!inLeviathanRegion() || (plugin.notified && !config.notifyRepeatedly()))
        {
            return;
        }
        if (!(hitsplatApplied.getActor() == client.getLocalPlayer()))
        {
            return;
        }

        Hitsplat hitsplat = hitsplatApplied.getHitsplat();

        // Check for Divine Potion hitsplats
        if (hitsplat.getAmount() == 10 && isDivinePotionActive())
        {
            boolean isFromAttack = checkCombatActivity();
            if (!isFromAttack)
            {
                log.debug("Ignored Divine Potion hitsplat of 10 HP.");
                return;
            }
        }

        // Check for Ruby Bolts (e) Blood Forfeit
        int playerHp = client.getLocalPlayer().getHealth();
        int expectedDamage = (int) Math.ceil(playerHp * 0.10); // 10% of current HP
        if (hitsplat.getAmount() == expectedDamage && isRubyBoltsEffectActive())
        {
            log.debug("Ignored Ruby Bolts (e) Blood Forfeit hitsplat.");
            return;
        }

        // Existing hitsplat handling logic
        if (hitsplat.isMine() && hitsplat.getHitsplatType() != HitsplatID.BLOCK_ME)
        {
            Deque<GraphicsObject> graphicsObjects = client.getGraphicsObjects();
            Iterator<GraphicsObject> iterator = graphicsObjects.iterator();
            while (iterator.hasNext())
            {
                GraphicsObject obj = iterator.next();
                if (ArrayUtils.contains(LEVIATHAN_BOULDER_SHADOWS, obj.getId()))
                {
                    if (checkCollision(obj))
                    {
                        log.debug("Found possible collision");
                        log.debug("ObjectID: " + obj.getId());
                        log.debug("Animation Frame: " + obj.getAnimationFrame());
                        if (obj.getAnimationFrame() < 3)
                        {
                            return;
                        }
                        plugin.notifyFailure(LEVIATHAN, "You were hit by falling rubble");
                        return;
                    }
                }
            }

            // Check pathfinder or special attack
            if (ArrayUtils.contains(LEVIATHAN_SPECIAL_ATTACK_ANIMATIONS, leviathanSpecial))
            {
                if (LEVIATHAN_SMOKE_BLAST_ANIM == leviathanSpecial)
                {
                    plugin.notifyFailure(LEVIATHAN, "You were hit by the smoke blast.");
                    return;
                }
                if (ArrayUtils.contains(LEVIATHAN_SMOKE_BLAST_BOULDERS_ANIM, leviathanSpecial))
                {
                    plugin.notifyFailure(LEVIATHAN, "You were hit by falling rubble.");
                    return;
                }
                if (ArrayUtils.contains(LEVIATHAN_LIGHTNING_ANIM, leviathanSpecial))
                {
                    plugin.notifyFailure(LEVIATHAN, "You were hit by lightning.");
                    return;
                }
            }

            List<NPC> npcs = client.getNpcs();
            for (NPC npc : npcs)
            {
                if (npc.getId() == NpcID.ABYSSAL_PATHFINDER)
                {
                    plugin.notifyFailure(LEVIATHAN, "You were hit outside the pathfinder.");
                    return;
                }
            }
        }
    }

    private boolean checkCollision(GraphicsObject obj)
    {
        LocalPoint localPoint = obj.getLocation();
        WorldPoint worldPoint = WorldPoint.fromLocal(client, localPoint);
        return worldPoint.equals(plugin.lastLocation);
    }

    // Checks if the Divine Potion effect is active
    private boolean isDivinePotionActive()
    {
        int divinePotionVarbit = client.getVarbitValue(DIVINE_POTION_VARBIT_ID); // Replace with actual Varbit ID
        return divinePotionVarbit > 0;
    }

    // Checks if the Ruby Bolts (e) Blood Forfeit effect is active
    private boolean isRubyBoltsEffectActive()
    {
        boolean soundEffectTriggered = client.getLastSoundEffect() == BLOOD_FORFEIT_SOUND_ID; // Replace with actual ID
        boolean graphicTriggered = client.getLocalPlayer().getGraphic() == BLOOD_FORFEIT_GRAPHIC_ID; // Replace with actual ID
        return soundEffectTriggered || graphicTriggered;
    }

    // Checks for combat activity to distinguish Divine Potion hitsplats from other hits
    private boolean checkCombatActivity()
    {
        IterableHashTable<ActorSpotAnim> spotAnims = client.getLocalPlayer().getSpotAnims();
        return spotAnims.iterator().hasNext(); // Returns true if combat-related graphics are present
    }
}`


Fairly new to this still learning programming but decided to see if I could give this a crack. 

but to explain.

Divine Potion: Filters out self-inflicted damage from Divine Potions by checking the Varbit state and ensuring it’s not part of an attack.

Ruby Bolts (e): Detects the Blood Forfeit effect using either a sound or a graphic ID. 

Helper Methods: Organized as private methods to keep the logic clean and reusable.

**You would still need to update** 

    // Placeholders for IDs (replace with actual IDs using DevTools)
    private static final int DIVINE_POTION_VARBIT_ID = 1234;
    private static final int BLOOD_FORFEIT_SOUND_ID = 5678;
    private static final int BLOOD_FORFEIT_GRAPHIC_ID = 91011;

with the IDs

Hope this helps, if you still actively look at this. Just done over 1000 kills at leviathan was something that bugged me so thought id do my best to make it better.
